src/openrct2/ride/Vehicle.cpp

```diff
@@ -2579,53 +2560,42 @@ static void vehicle_update_waiting_to_depart(rct_vehicle* vehicle)
         {
             if (track.element->AsTrack()->HasCableLift())
             {
-                vehicle->status = VEHICLE_STATUS_WAITING_FOR_CABLE_LIFT;
+                vehicle->SetState(VEHICLE_STATUS_WAITING_FOR_CABLE_LIFT, vehicle->sub_state);
             }
         }
     }
 
     switch (ride->mode)
     {
         case RIDE_MODE_BUMPERCAR:
-            vehicle->status = VEHICLE_STATUS_TRAVELLING_DODGEMS;
-            vehicle_invalidate_window(vehicle);
             // Bumper mode uses sub_state / var_CE to tell how long
             // the vehicle has been ridden.
-            vehicle->sub_state = 0;
+            vehicle->SetState(VEHICLE_STATUS_TRAVELLING_DODGEMS, 0);
             vehicle->var_CE = 0;
             vehicle_update_dodgems_mode(vehicle);
             break;
         case RIDE_MODE_SWING:
-            vehicle->status = VEHICLE_STATUS_SWINGING;
-            vehicle->sub_state = 0;
-            vehicle_invalidate_window(vehicle);
+            vehicle->SetState(VEHICLE_STATUS_SWINGING, 0);
             vehicle->var_CE = 0;
             vehicle->current_time = -1;
             vehicle_update_swinging(vehicle);
             break;
         case RIDE_MODE_ROTATION:
-            vehicle->status = VEHICLE_STATUS_ROTATING;
-            vehicle->sub_state = 0;
-            vehicle_invalidate_window(vehicle);
+            vehicle->SetState(VEHICLE_STATUS_ROTATING, 0);
             vehicle->var_CE = 0;
             vehicle->current_time = -1;
             vehicle_update_rotating(vehicle);
             break;
         case RIDE_MODE_FILM_AVENGING_AVIATORS:
         case RIDE_MODE_FILM_THRILL_RIDERS:
-            vehicle->status = VEHICLE_STATUS_SIMULATOR_OPERATING;
-            vehicle->sub_state = 0;
-            if (ride->mode == RIDE_MODE_FILM_THRILL_RIDERS)
-                vehicle->sub_state = 1;
-            vehicle_invalidate_window(vehicle);
+            vehicle->SetState(VEHICLE_STATUS_SIMULATOR_OPERATING, 1);
```

Comment:
This isn't quite right

Reply:
In the original code , vehicle->substate will always be set to 1, because `case RIDE_MODE_FILM_THRILL_RIDERS` and `if (ride->mode == RIDE_MODE_FILM_THRILL_RIDERS)` appear to be redundant. Unless I've missed something.

Reply:
But what about RIDE_MODE_FILM_AVENGING_AVIATORS

Reply:
Ah yes, I see now. I gave it's own case, for the sake of explicitness.